### DISCUSSION TITLE
Customising filetypes

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/onedarkpro.nvim#98] [Announcements]
Hey @amaanq. Moving the discussion over here.

I have filetype highlights set out like this:

```lua
    highlights = {
      -- Ruby
      ["@function.ruby"] = { fg = "${blue}", style = "bold" },
      ["@function.call.ruby"] = { fg = "${blue}", style = "bold" },
      ["@include.ruby"] = { fg = "${blue}", style = "italic" },
      ["@parameter.ruby"] = { fg = "${fg}", style = "italic" },
      ["@symbol.ruby"] = { fg = "${cyan}" },

      -- Yaml
      ["@field.yaml"] = { fg = "${red}" },
      ["@string.yaml"] = { fg = "${green}" },
  }
```

Fancy trying for Rust?

**Comment by amaanq** [comment]
oh that's neat for specific languages, but Rust has a lot of unusual operators, for example the try_expression I mentioned earlier is a ? at the end of a result or option, how would that be customized in Rust? 

https://github.com/tree-sitter/tree-sitter-rust here's the Rust grammar for tree sitter if you want to maybe take a look, I want to really separate the components like how vscode has done it, I really liked it

**Reply by olimorris** [reply]
Send me a rust code snippet and a screenshot of how it looks in Visual Studio Code and I'll have a play

**Reply by amaanq** [reply]
sorry, just saw this!

![image](https://user-images.githubusercontent.com/29718261/193719177-8b71aa0e-e801-42a5-b443-c2ad4b278594.png)
here you can see the ? and & operator are white, and the Some(_) expression is cyan. That's a smaller example to see if you can maybe find what the corresponding tree-sitter highlight group name is :)

**Reply by amaanq** [reply]
ok, adding on since I found a theme that uses a lot of rust-specific highlight groups, but not all of them. I just wish I knew where these could be easily/readily found... https://github.com/lmburns/kimbox/blob/22f37d5a217d65af1aa8c742e67d534340b2fdb1/lua/kimbox/highlights.lua#L1072

from that, ive made some adjustments of my own and I am pleased, but theres still many left to do.. I'll edit this as I keep going, but something I've found odd is the style = "italic" fields dont work at all

```lua
highlights = {
    ["@field.rust"] = { fg = "${red}" },
    ["@function.macro.rust"] = { fg = "${orange}" },
    ["@keyword.rust"] = { fg = "${purple}" },
    ["@label.rust"] = { fg = "${white}" },
    ["@operator.rust"] = { fg = "${white}" },
    ["@parameter.rust"] = { fg = "${red}", style = "italic" },
    ["@punctuation.bracket.rust"] = { fg = "${purple}" },
    ["@variable.builtin.rust"] = { fg = "${purple}", style = "italic" },
}
```

Adding a very clear example of how vscode differs from this theme in neovim, and how neovim feels off/harder to read because of this (Top is vscode, bottom is neovim)
![image](https://user-images.githubusercontent.com/29718261/193762650-bc6b5b82-94f9-4338-bd69-d8dc28ba21c9.png)
![image](https://user-images.githubusercontent.com/29718261/193762674-0ad7e2a1-09dd-4676-8268-8cdbfa4577dc.png)

**Reply by olimorris** [reply]
I think there is a `@function.builtin.rust` which can be set to `cyan` to make it _really_ close.

Italics are odd. I shall invesigate!

**Reply by amaanq** [reply]
thanks! it was actually @constant.builtin.rust, but that fixed the Some, None, Ok, Err colors! do you happen to know how to color enum items? 
```rust
enum Color {
    Red,
    Blue,
    Green,
}
```
normally, enum items are orange but I'd like them to be cyan (and thats the default in vscode)

**Reply by olimorris** [reply]
With these

```lua
      ["@operator.rust"] = { fg= "${fg}" },
      ["@punctuation.bracket.rust"] = { fg= "${yellow}" },
      ["@constant.builtin.rust"] = { fg= "${cyan}" },
```

I can get to this:

<img width="976" alt="Screen Shot 2022-10-04 at 08 32 25@2x" src="https://user-images.githubusercontent.com/9512444/193760911-8f57f400-e5e4-4572-984e-e62731800e26.png">

EDIT: Sorry this didn't post earlier

**Reply by olimorris** [reply]
> thanks! it was actually @constant.builtin.rust, but that fixed the Some, None, Ok, Err colors! do you happen to know how to color enum items?
> 
> ```rust
> enum Color {
>     Red,
>     Blue,
>     Green,
> }
> ```
> 
> normally, enum items are orange but I'd like them to be cyan (and thats the default in vscode)

I think we're hitting the limit of what is exposed via Treesitter:

<img width="359" alt="Screen Shot 2022-10-04 at 11 05 49@2x" src="https://user-images.githubusercontent.com/9512444/193792945-167e45b6-a4f8-4565-9155-602d02db8313.png">

Can't seem to find anything specifically related to enums

**Reply by amaanq** [reply]
Ahh I see, that's really unfortunate. Thank you for the help so far!

**Reply by olimorris** [reply]
@amaanq could you share your final Rust configuration please? I'd like to incorporate into the colorscheme

**Reply by amaanq** [reply]
Sorry for getting back so late, I hadn't noticed your message until I checked back here again! Here's my current Rust config!

I also added a toml config as by default properties are purple in vscode's onedarkpro theme, might be nice to add too

```lua
highlights = {
    ["@constant.builtin.rust"] = { fg = "${cyan}" },
    ["@field.rust"] = { fg = "${red}" },
    ["@function.builtin.rust"] = { fg = "${cyan}" },
    ["@function.macro.rust"] = { fg = "${orange}" },
    ["@keyword.rust"] = { fg = "${purple}" },
    ["@label.rust"] = { fg = "${white}" },
    ["@operator.rust"] = { fg = "${fg}" },
    ["@parameter.rust"] = { fg = "${red}", style = "italic" },

    ["@property.toml"] = { fg = "${purple}" },
}
```

I can also upload a golang config within a week or so, just a bit busy atm and I haven't touched go in a bit

While we're on the topic of highlights/theming, something that has been irking me is how a diagnostic warn/error will color the text with the type (red for err, yellow for warn, etc). I don't particularly like this and **only** want the undercurls. I noticed setting undercurl to true in the options does add the undercurls, but it does not turn off this text highlighting, here's an example:

![image](https://user-images.githubusercontent.com/29718261/194715986-e2eb6882-f446-4e8e-8f02-a554fb540302.png)

Although it is somewhat of a false positive warning, the matter at hand is I just don't want capabilities (on the left) to be yellow in color, I want it simply to have a yellow undercurl. Let me know how to fix this, thanks!

**Comment by olimorris** [comment]
@mmirus and @mrjones2014 - Are you using the original filetype highlights at all? I'm looking to deprecate them soon in favour of the "new" treesitter filetype highlights.

**Reply by mrjones2014** [reply]
Nope, the only thing related to a specific filetype I have is these under the normal `highlights` config key:

```lua
        ['@keyword.operator.lua'] = { fg = '#d55fde', style = 'italic' },
        ['@operator.lua'] = { link = '@keyword.operator.lua' }
```

**Reply by mmirus** [reply]
Nope! Only the new syntax in `highlights` now. Thanks for asking.

**Reply by olimorris** [reply]
Super. I'll put a deprecation notice up and warn people and remove in a week or two. Thanks both.
